### DISCUSSION TITLE
The recipe for graphql-ws client with apollo is not working with Vanilla JS

### DISCUSSION THREAD

**Discussion by NicoSan20** [enisdenjo/graphql-ws#300] [Q&A] [answered]
Good morning

As mentioned in the title, the recipe for graphql-ws client with apollo is not working with Vanilla JS. This example was made for TypeScript.

Could you tell me how to adapt this example to work with Vanilla JS please ?

Thank you in advance for your help.

**Comment by enisdenjo** [comment, marked as answer]
[A simple Google search with "typescript to javascript"](https://www.google.com/search?channel=nrow5&client=firefox-b-d&q=typescript+to+javascript) gives you the tools. [TS Playground from the results](https://www.typescriptlang.org/play?#code/PTAEDMHsCdQQQA6QDbMqAwsglgUwHYAuoAbgMwBcAUNgLZLTEDeVo8SqkAMtvgNYAaVqADyCXNACGhbJHxC2AMVyEAxgAsASrgDOAV2SEFogEY6JJSSeS4hAXwjRItUAHIAApI5pgqnAUJfGFxXAG4qEFAYCGjEFDRMfyJSACZqSLoGZnZ47l5BUXEpGTkBUGU1LV0DI1NzaEtrXFAHcCcXVy9cgFocfjDQSIBPSWh8UEkAE0mJ70he-KoaehhshGheYlb2twBzKQR1AEdkMOWs0CZQVWhcaVwsPCIyx4CxEvwdF6Stx2c9g7HZDdADuOjOVD8kh0OlAAHVcCYAMqQVR8FQ8figXAAD0IBEmsLinExfEuwnW2Es+OuPwoiSehHCwlUch0hGgelUhBgAApIAgPjp6a8iO9ZJ8AJTkthsfRFXmS8Ky0CEdTYHQAOj8jNAAF5rrd7qLCPzBRKdErhHYlmwEHprNhVKBbkc9LpTQKJNIJfSxN6PpK-WYLFYbAAeCoabT6QwAPhlstuhD0Y1A+FwILqoaavN5Ony0r1CZYKrYydT4zVGu1P01+jMN2wJlwkZU0eq8d5wjLl1AmoHXuKErKbokQ3plKIZoDEs1Y+gQ2ldmMvdLvdlGbx9IL-E1W8ImpMvEm+cLq43rPoNnxO-y2ucCBvuCPJ7P-ElF97Eic0Dve5-GBX3wU9dz4T8ezLFdILYK0VTsOCWioG1ITZYg+jJA0MyzBFkVRdFCFJXl11TZB6VcMEKBAEF1AkF8NQoAAWAAGVjgH2LwgVcYxWXwDNuQlAAFUZJFoYVQEVfUS0g3j2VAcwYQlfVQF2FQkV0As5EVZVZWwcAJIAQgUzT8GldcVQrNMmDsHS2BtCyVErRMVTgPQ1RgbAAC8fTkekAAMACE7luWAABImGMuceXRfA7D8i8bOtewlSAA) would give you a JavaScript only version like this:
```js
// for Apollo Client v3:
import { ApolloLink, Observable, } from '@apollo/client/core';
// or for Apollo Client v2:
// import { ApolloLink, Operation, FetchResult, Observable } from 'apollo-link'; // yarn add apollo-link
import { print } from 'graphql';
import { createClient } from 'graphql-ws';
class WebSocketLink extends ApolloLink {
    constructor(options) {
        super();
        this.client = createClient(options);
    }
    request(operation) {
        return new Observable((sink) => {
            return this.client.subscribe(Object.assign(Object.assign({}, operation), { query: print(operation.query) }), {
                next: sink.next.bind(sink),
                complete: sink.complete.bind(sink),
                error: sink.error.bind(sink),
            });
        });
    }
}
const link = new WebSocketLink({
    url: 'ws://where.is:4000/graphql',
    connectionParams: () => {
        const session = getSession();
        if (!session) {
            return {};
        }
        return {
            Authorization: `Bearer ${session.token}`,
        };
    },
});
```

P.S. you should learn TypeScript ;)

**Reply by NicoSan20** [reply]
Hi @enisdenjo ! Thank you so much for your reply and for your time. Really appreciate it. This work fine. Again, thank you... Best regards
